Feature/aos 3121 bytte av response format til hal hypermedia format

kristofferchr opened the pull request:
Gjenstår refaktorering

kristofferchr committed:
Replaced "-" with "_" for Docker content digest manifest

kristofferchr committed:
Replaced RestTemplate with WebClient
Changed port and debugging in application.yaml to make it possible to test against Gobo

kristofferchr committed:
Implemented detection of manifest version support for response
Better error handling
Problem with blocking

kristofferchr committed:
Working implementation of both V1 and V2 support

kristofferchr committed:
Started implementation of HAL as response format

kristofferchr committed:
Refactoring
Added remanining uri template

kristofferchr committed:
404 error from Docker registry now returns empty map or list
Disabled test that is not applicable anymore

kristofferchr committed:
refactoring

kristofferchr committed:
Readme update

kristofferchr committed:
Refactoring

kristofferchr committed:
Refactoring

kristofferchr committed:
Merge branch 'feature/AOS-3109-stotte-for-manifest-skjema-v2-og-v1-uthenting' of https://github.com/Skatteetaten/cantus into feature/AOS-3121-bytte-av-response-format-til-hal-hypermedia-format

kristofferchr committed:
Started to implement service layer DTOs and Controller layer dtos

kristofferchr committed:
Able to get manifest, tags and grouped tags with HAL format as response

kristofferchr committed:
Merge remote-tracking branch 'origin/master' into feature/AOS-3121-bytte-av-response-format-til-hal-hypermedia-format

bjartek commented:
Idag trekkes det inn et hal bibliotek her, men det er ikke under aktiv utvikling. Skulle vi laget et eget lite aurora-kotlin-utils bibliotek med kode som vi gjenbruker veldig mye? Alternativt skal vi lage mange små?

Et bibliotek med exceptions og responseklasser og litt extension methods kanskje?

- https://github.com/davidsowerby/hal-kotlin

bjartek commented on src/main/kotlin/no/skatteetaten/aurora/cantus/service/DockerRegistryService.kt:
```diff
@@ -142,22 +100,60 @@ class DockerRegistryService(
+            imageManifestResponse.dockerContentDigest == null ||
+            imageManifestResponse.manifestBody == null
+        ) {
+            return ImageManifestDto(dockerDigest = "", dockerVersion = "")
```
Mister vi ikke noe informasjon her? Burde vi her hive en exception?

kristofferchr replied:
Det blir nå hivet en exception av type SourceSystemException som blir tatt hånd om av Error handleren. Responsen som kommer til kalleren er da et response av type HAL format med exception inni denne.

kristofferchr committed:
Better error handling

kristofferchr committed:
Continued work on error handling

kristofferchr committed:
Updated tests for the new response format
Added more tests

kristofferchr committed:
Added test for V2 schema

jarlehansen commented on src/main/kotlin/no/skatteetaten/aurora/cantus/controller/DockerRegistryController.kt:
```diff
@@ -15,23 +18,87 @@ class DockerRegistryController(val dockerRegistryService: DockerRegistryService)
+
+@Component
+class ImageTagResoureAssembler {
+    companion object {
```
Denne er både en @Component og har companion object. Virker litt unødvendig for meg, er det en grunn til det?

bjartek replied:
Det var meg som skrev dette og jeg tror bare jeg har surret litt. Hva er det vi gjør andre steder? Så lenge det er likt som i mokey og de andre er det samme for meg hva vi velger.

jarlehansen replied:
@Component er nok å foretrekke, pga enklere mocking

kristofferchr replied:
Companion object er fjernet og bean er laget for ImageTagResourceAssembler

jarlehansen commented on src/main/kotlin/no/skatteetaten/aurora/cantus/controller/ErrorHandler.kt:
```diff
@@ -44,15 +49,15 @@ fun <T> Mono<T>.blockAndHandleError(duration: Duration = Duration.ofSeconds(30),
 private fun <T> Mono<T>.handleError(sourceSystem: String?) =
     this.doOnError {
-
+        println(it.message)
```
litt debugging kode som henger igjen :)

kristofferchr replied:
Se der har det sneket seg med en sånn :) Fikser det!

jarlehansen commented on src/main/kotlin/no/skatteetaten/aurora/cantus/service/DockerRegistryService.kt:
```diff
@@ -68,95 +68,99 @@ class DockerRegistryService(
+        } ?: throw SourceSystemException(
+            "Manifest not found for image $imageGroup/$imageName:$imageTag",
+            code = "404",
+            sourceSystem = url
```
Source system er ment for å vise system navnet der den opprinnelige feilen skjer, kanskje det burde stå "docker registry" her?

bjartek replied:
Poenget her er at vi har flere docker registies vi går mot. Vi må ha en måte å skille dem på. Hvis vi får masse feil fra interne registry men ikke eksterne?

jarlehansen replied:
Godt poeng, bruk det som gir oss mest verdi når vi trenger å feilsøke. hvis det er url'en så bare behold det.

kristofferchr replied:
Kan det her være lurt å sette på en prefix for urlen eller vet man at det er Docker registry ut ifra alle urlene som kan brukes?

jarlehansen commented on src/test/kotlin/no/skatteetaten/aurora/cantus/controller/DockerRegistryControllerTest.kt:
```diff
@@ -32,34 +46,98 @@ class DockerRegistryControllerTest {
+    fun `Get docker registry image info given missing resource`(path: String) {
+        val dockerUrl = "https://localhost.no"
+
+        `when`(dockerService.getImageTags(any(), any(), anyOrNull())).thenThrow(
```
Siden when er et reservert ord i kotlin så må man bruke `
Jeg foretrekker heller å bruke BDDMockito.given()

kristofferchr replied:
Så when i java er det samme som given i kotlin? :) Takk!

jarlehansen replied:
Nei, ikke helt. "when" i Kotlin er et reservert ord, det er det ikke i Java. Så når man skal lage en funksjon i Kotlin som heter when så må den skrives `when` sånn. Så får å unngå denne syntaxen, så foretrekker jeg å bruke given. Given gjør det samme som when i mockito, bare at det er BDD måten å sette det opp på.

kristofferchr replied:
Jeg skjønner :) Fikset nå.

kristofferchr committed:
Cleaning
Replaced `when` with given in test
Removed companion object from ImageTagResourceAssembler and added corresponding bean

kristofferchr committed:
Refactoring

kristofferchr committed:
Cleaning

Commit:
fixed jenkinsfile
